src/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Sockets.Tests
+{
+    public class NetworkStreamTest
+    {
+        [Theory]
+        [MemberData("NonCanceledTokens")]
+        public async Task ReadWriteAsync_NonCanceled_Success(CancellationToken nonCanceledToken)
+        {
+            await RunWithConnectedNetworkStreamsAsync(async (server, client) =>
+            {
+                var clientData = new byte[] { 42 };
+                await client.WriteAsync(clientData, 0, clientData.Length, nonCanceledToken);
+
+                var serverData = new byte[clientData.Length];
+                Assert.Equal(serverData.Length, await server.ReadAsync(serverData, 0, serverData.Length, nonCanceledToken));
+
+                Assert.Equal(clientData, serverData);
+            });
+        }
+
+        [Fact]
+        public async Task ReadWriteAsync_Canceled_ThrowsOperationCanceledException()
```

Comment:
> LGTM for merge

Thanks.

>  I still don't understand which one APIs in general are supposed to throw and how they decide on which one

If we had it to do over again, we probably wouldn't have introduced TaskCanceledException at all.  Code outside of the TPL implementation is unlikely to ever throw a TaskCanceledException; the point of the exception was that, in cases where TPL was manufacturing the exception, TPL could pass with the OperationCanceledException information about which Task was canceled, in case it was useful to the consumer, who might do a type check and then query the Task property.  That property is very rarely used or needed.